### ISSUE TITLE
Reduce CI time

### ISSUE THREAD

**Issue by zanieb** [astral-sh/uv#5713]
Running tests in CI is now >8 minutes (and was previously <1 minute). We've done a lot to optimize this previously (TODO: add references), but I'd like to start discussing how we can improve this overall.

See also:

- https://github.com/astral-sh/uv/issues/5711
- #878

**Comment by zanieb** [comment]
Would it be absurd to introduce "CI" time regression checks in CI like the CodSpeed benches? Unfortunately the GitHub Runner performance is super noisy so it might not work.

**Comment by helderco** [comment]
Wonder if [Dagger](https://dagger.io) could help here 🙂

**Comment by samypr100** [comment]
Arguably self-hosted runners could possibly help here too, but the maintenance/cost burden is likely too large

**Comment by zanieb** [comment]
I strongly considered self-hosted runners, but it seemed painful to orchestrate Windows runners in particular.

**Comment by zanieb** [comment]
I also very much want to look into something like https://github.com/astral-sh/uv/pull/609 again to cache our network traffic — I think that'd help a lot.

**Comment by samypr100** [comment]
> I also very much want to look into something like #609 again to cache our network traffic — I think that'd help a lot.

At least for python, maybe worth using something like https://github.com/hauntsaninja/nginx_pypi_cache and saving a copy of the cache as a github cache and reloading it. I actually use it locally a lot to speed up tests.

**Comment by charliermarsh** [comment]
That strikes me as a good idea...

**Comment by zanieb** [comment]
Ah that might be easier than using mitmproxy or rolling our own proxy in Rust. Thanks for the link!

**Comment by zanieb** [comment]
Some progress with:

- #5874 
- #5883 
- #5873

**Comment by zanieb** [comment]
Linux is acceptably fast now. We're at the limit for macOS machine size without going to alternative runner providers. There are larger Windows runners, maybe I should test one. Cost may be a problem at some point, alternative runner providers may be cheaper (often saying things like a 2x cost reduction).

**Comment by ChannyClaus** [comment]
have we considered using https://bazel.build/? (correct me if i'm wrong but `cargo test` doesn't seem to skip the tests when none of their dependencies changed)

admittedly adding bazel would significantly increase the complexity though 🙀

**Comment by eth3lbert** [comment]
Yes, I've heard a lot about complaints regarding bazel's complexity.

**Comment by ChannyClaus** [comment]
> Yes, I've heard a lot about complaints regarding bazel's complexity.

this is definitely true - i've seen it improving the CI time drastically at the same time though so maybe something to consider down the line if there's no other option...
